### PR TITLE
Add a Network Folder Canarytoken

### PR DESCRIPTION
**Pull request by thinkst-marco**
## Proposed changes

This PR introduces a significant chunk of work originally by Jacob Torrey, to support a new Network Folder Canarytoken.

Based on a lightweight WebDAV server implemented as a Cloudflare worker, users will map a long-lived connection to a WebDAV server that we monitor for access. When the user's unique credentials are used to open files, the Canarytoken will fire.

## Types of changes

- [X] New feature (non-breaking change which adds functionality)

## Checklist

- [X] Lint and unit tests pass locally with my changes (if applicable)
- [X] I have run pre-commit (`pre-commit` in the repo)
- [X] I have added tests that prove my fix is effective or that my feature works

## Further comments

The failing unit test is unrelated to this change.

### CODE REVIEW

File: tests/units/test_frontend.py [hunk 1]
```diff
@@ -716,6 +720,129 @@ def test_aws_keys(
...
+        "nosuchaccountid",
+        "nosuchapitoken",
+        "nosuchnamespace",
+    )
+    overrides = {
+        "WEBDAV_SERVER": webdav_server,
+        "CLOUDFLARE_ACCOUNT_ID": cf_account_id,
+        "CLOUDFLARE_API_TOKEN": cf_api_token,
+        "CLOUDFLARE_NAMESPACE": cf_namespace,
+    }
```
Comment (gjcthinkst) [inline tests/units/test_frontend.py:739]:
```suggestion
    overrides = {
        "WEBDAV_SERVER": "https://examplewebdavserver.com",
        "CLOUDFLARE_ACCOUNT_ID": "nosuchaccountid",
        "CLOUDFLARE_API_TOKEN": "nosuchapitoken",
        "CLOUDFLARE_NAMESPACE": "nosuchnamespace",
    }
```
Nitpick

File: tests/units/test_frontend.py [hunk 2]
```diff
@@ -716,6 +720,129 @@ def test_aws_keys(
...
+        assert token_info.webdav_fs_type
+        assert token_info.webdav_password
+        assert token_info.webdav_server
+
+
+def test_webdav_no_cloudflare(
+    webhook_receiver: HttpUrl,
+    frontend_settings: FrontendSettings,
+) -> None:
+    (webdav_server, cf_account_id, cf_api_token, cf_namespace) = ("", "", "", "")
```
Comment (gjcthinkst) [inline tests/units/test_frontend.py:793]: We can also rm this similar to the above suggestion.

File: tests/units/test_frontend.py [hunk 3]
```diff
@@ -716,6 +720,129 @@ def test_aws_keys(
...
+            resp = _create_webdav_token_response(
+                token_request_details=token_request_details,
+                canarydrop=cd,
+                settings=local_settings,
+            )
+
+        token_info = WebDavTokenResponse(**resp.dict())
+        assert token_info.webdav_fs_type
+        assert token_info.webdav_password
+        assert token_info.webdav_server
```
Comment (gjcthinkst) [inline tests/units/test_frontend.py:786]: It's not clear to me what we're trying to test here. Are we just testing that the response is valid according to the schema? Can we also test that the fs type is correct (`"it"` in this case), and that password and server are valid, or at least have the correct shape.

A docstring might also make it more clear what we're trying to test here.

File: tests/units/test_frontend.py [hunk 4]
```diff
@@ -716,6 +720,129 @@ def test_aws_keys(
...
+    }
+    with mock.patch.dict(
+        os.environ,
+        {"CANARY_" + k: v for (k, v) in overrides.items()},
+        clear=False,
+    ):
+        local_settings = frontend_settings.dict()
+        for k, v in overrides.items():
+            local_settings[k] = v
+        local_settings = FrontendSettings(**local_settings)
```
Comment (gjcthinkst) [inline tests/units/test_frontend.py:808]: Can we extract this into a fixture to remove a level of indentation and keep the setup code out of the test itself?

File: frontend/app.py [hunk 5]
```diff
@@ -1604,6 +1607,56 @@ def _create_azure_id_token_response(
...
+    ):
+        return JSONResponse(
+            {
+                "error_message": "This Canarytokens instance does not have the Network Folder Canarytoken enabled."
+            },
+            status_code=400,
+        )
+    canarydrop.webdav_fs_type = FsType(token_request_details.webdav_fs_type)
+    canarydrop.webdav_server = settings.WEBDAV_SERVER
+    queries.save_canarydrop(canarydrop=canarydrop)
```
Comment (gjcthinkst) [inline frontend/app.py:1634]: Do we need to save before generating the password (doesn't look like the password generation can really fail)?

**Review by gjcthinkst — COMMENTED** [review]
I see the latest run of the unit tests failed. Is that expected?

File: tests/integration/test_webdav.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    )
+    assert len(webdav.ls()) > 0
+
+    webdav_token_fire(token_info)
+
+    token_hist_resp = get_token_history(token_info=token_info, version=v3)
+
+    token_hist = WebDavTokenHistory(**token_hist_resp)
+
+    assert len(token_hist.hits) == 1
```
[the remark this answers is not in the document]
Comment (thinkst-marco) [reply tests/integration/test_webdav.py:73]: The integration test uses a temporary private Canarytokens instance that isn't accessible by Cloudflare, so the trigger in line 59 will never reach the integration test's instance. Hence the need for the manual trigger on line 67.

To be sure about what this test tells us: the test confirms that we can browse the Cloudflare-hosted WebDAV server, and that the Canarytokens server can process incoming reports from Cloudflare. But it cannot test that the Cloudflare worker sends the alert in the right format.

File: tests/units/test_frontend.py (hunk 4 above)
Comment (thinkst-marco) [reply tests/units/test_frontend.py:802]: Extracted into a fixture.

**Comment by thinkst-marco** [comment]
The failed unit tests pass now (https://github.com/thinkst/canarytokens/actions/runs/12118441400/job/33782929162) , without making any changes to that code. They appear to be unrelated to this work. The new failures related to the incoming FakeFS token, so I'm ignoring them.

File: tests/integration/test_webdav.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gjcthinkst) [reply tests/integration/test_webdav.py:73]: That makes sense, thanks!

**Review by gjcthinkst — APPROVED** [review]
> The failed unit tests pass now (https://github.com/thinkst/canarytokens/actions/runs/12118441400/job/33782929162) , without making any changes to that code. They appear to be unrelated to this work. The new failures related to the incoming FakeFS token, so I'm ignoring them.

Sounds reasonable.

**Pull request merged**